Get better cost estimate on MultiTermQuery over few terms

msfroh opened the pull request:
### Description

The existing cost estimate for multi-term queries was taking the worst-case scenario (e.g the query matches every term in the field), to provide an upper bound on the cost.

In cases where the MultiTermQuery matches a small number of terms, we can provide a tighter estimate by summing the doc frequencies over those terms.

Fixes #13029

msfroh committed:
Get better cost estimate on MultiTermQuery over few terms

The existing cost estimate for multi-term queries was taking the
worst-case scenario (e.g the query matches every term in the field),
to provide an upper bound on the cost.

In cases where the MultiTermQuery matches a small number of terms, we
can provide a tighter estimate by summing the doc frequencies over
those terms.

msfroh commented on lucene/core/src/java/org/apache/lucene/search/AbstractMultiTermQueryConstantScoreWrapper.java:
```diff
@@ -292,7 +292,21 @@ public long cost() {
+    private long estimateCost(Terms terms, long queryTermsCount) throws IOException {
+
+      // Try to explicitly compute the cost, but only if the term count is sufficiently small.
+      if (queryTermsCount < MAX_TERMS_TO_COUNT) {
```
Correct -- I thought about making that more explicit, but opted to take advantage of the "fall through" logic.

msfroh commented on lucene/core/src/java/org/apache/lucene/search/AbstractMultiTermQueryConstantScoreWrapper.java:
```diff
@@ -292,7 +292,21 @@ public long cost() {
     }
 
-    private static long estimateCost(Terms terms, long queryTermsCount) throws IOException {
+    private static final int MAX_TERMS_TO_COUNT = 128;
```
It's a great question! Maybe we could pass it as a constructor parameter, and the constant `RewriteMethod`s in `MultiTermQuery` could specify the default?

I host a weekly OpenSearch Lucene Study Group on Zoom and we spent this week's meetup talking about this PR (and your issue): https://forum.opensearch.org/t/opensearch-lucene-study-group-meeting-monday-march-25th-2024/18547/3

There were some nice ideas that came up there around how to pick something better than an arbitrary limit -- like maybe using a time threshold instead. (But would that be a threshold per-clause?)

rquesada-tibco replied:
That was indeed an interesting discussion. Thanks for sharing @msfroh !

msfroh commented:
I should check if maybe this has been superseded by https://github.com/apache/lucene/pull/13454

That change seems to look at some of the term states for the first 16 terms during rewrite. Maybe that improves the cost estimate?

Or maybe this change can work on the assumption that the first 16 terms are available and use that to produce a better cost estimate if those (up-to) 16 terms are exhaustive.

msfroh commented:
There we go -- I reworked the PR to reuse the terms collected during the call to `rewrite` in https://github.com/apache/lucene/pull/13454.

Rather than modifying `estimateCost`, we can say "If the MTQ will rewrite to a BQ, then use the BQ's cost".

msfroh commented on lucene/core/src/java/org/apache/lucene/search/AbstractMultiTermQueryConstantScoreWrapper.java:
```diff
@@ -154,21 +154,24 @@ protected abstract WeightOrDocIdSetIterator rewriteInner(
+        LeafReaderContext context,
+        Terms terms,
+        List<TermAndState> collectedTerms,
+        boolean[] collectResult)
```
Turning `collectResult` into an out-param (along with `collectedTerms`) feels hacky.

I might try inlining `rewrite` (folding it into `scorerSupplier`) to see how that looks.

msfroh committed:
Inline rewrite method

I needed to share so many variables between scorerSupplier and rewrite
that I think it makes more sense to just inline rewrite so that the
variables are all in scorerSupplier.

msfroh commented:
@gsmiller, @mayya-sharipova -- What do you both think of this approach? Thanks!

gsmiller commented on lucene/core/src/java/org/apache/lucene/search/AbstractMultiTermQueryConstantScoreWrapper.java:
```diff
@@ -240,9 +219,38 @@ public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOExcepti
+      List<TermAndState> collectedTerms = new ArrayList<>();
+      boolean collectResult = collectTerms(fieldDocCount, termsEnum, collectedTerms);
+
+      if (collectResult && collectedTerms.isEmpty()) return null;
```
nit: stylistically at least, I think we generally prefer always using braces even for one-liners. Mind including them here?

gsmiller commented on lucene/core/src/java/org/apache/lucene/search/AbstractMultiTermQueryConstantScoreWrapper.java:
```diff
@@ -240,9 +219,38 @@ public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOExcepti
+      if (collectResult && collectedTerms.isEmpty()) return null;
+
+      final long cost;
+      if (collectResult) {
```
We end up checking `collectResult` twice in the common case here. We could refactor this slightly to have one conditional block for when `collectResult == true` instead (and pull the `isEmpty` short-circuit check into it). WDYT?

gsmiller commented on lucene/core/src/java/org/apache/lucene/search/AbstractMultiTermQueryConstantScoreWrapper.java:
```diff
@@ -240,9 +219,38 @@ public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOExcepti
+      if (collectResult) {
+        long sumTermCost = 0;
+        for (TermAndState collectedTerm : collectedTerms) {
+          sumTermCost += collectedTerm.docFreq;
```
Just sort of thinking out loud here, but I wonder how expensive it is to build the boolean query and grab its weight / scoreSupplier at this point (instead of lazily doing it in `scoreSupplier#get`)? If it's cheap, another option would be to invoke `rewriteAsBooleanQuery`, pull its score supplier and then delegate `#cost` to it. That way if the cost logic in `BooleanQuery` evolves in the future, this benefits as well. Probably not worth messing with this now, but maybe worth a TODO comment in here? That is, unless we know it is prohibitively expensive.

gsmiller reviewed (commented):
Left a couple very minor comments and thoughts. Thanks @msfroh ! Looks good.

Commit:
small tweaks

Commit:
changes

gsmiller committed:
Merge branch 'main' into sum_multiterm_doc_freqs

gsmiller commented:
@msfroh hope you don't mind, but since my PR feedback was pretty minor, I went ahead and made those small changes on your branch and merged. I'll work on getting this backported shortly (a few conflicts to resolve). If you spot anything in the change you disagree with, please let me know and we can work through that. Thanks again!

msfroh commented:
Thanks, Greg! I can create a follow-up PR with your suggestion on delegating to the rewritten `BooleanQuery` for the cost estimate.

I think if we go down that path (basically doing the rewrite early), we might be able to clean up the code a bit. I think `WeightOrDocIdSetIterator` might become unnecessary (since we'll know ahead of time which one we'll have).

gsmiller commented:
Thanks @msfroh. Yeah, I think it's worth exploring as long as there isn't something expensive-ish hiding in the weight creation of `BooleanQuery`. Thanks again!
